### ISSUE TITLE
Calling Expired Progressive Calls

### ISSUE THREAD

**Issue by ecorm** [wamp-proto/wamp-proto#448]
What happens when a CALL message uses a request ID for a progressive call that no longer exists (due to the callee unregistering or leaving) and the caller has not yet received the `wamp.error.cancelled` error? I don't think aborting with `wamp.error.protocol_violation` would be appropriate here, because the race condition is outside the caller's control.

I propose that in such scenarios, the dealer responds with an error having a `wamp.error.no_such_call` URI. `wamp.error.no_such_procedure` would not be appropriate here because the procedure may still exist - it's merely the progressive call that has expired.

Note that when progressive calls are supported, a router can no longer simplistically reject CALL request IDs that are not sequential. The router must allow the caller to use an old request ID that may reference a recently expired progressive call. It would be too costly for a router to remember all previous progressive call request IDs.

There is therefore a scenario where a CALL uses an old request ID for a request that is still active, but is not a progressive call. In those instances, I propose that they be treated as protocol violations.

(This issue is for the continuation of a discussion I started in PR #445)

**Comment by KSDaemon** [comment]
Definitely, a situation when a router receives a call with a request ID of an ongoing non-progressive call should be treated as `protocol_violation`. That can not happen in a legitime flow.

The case when a router receives a progressive call with a request ID that doesn't exists in the router runtime can be treated as a starting new progressive call. The only way to distinguish whether is it an obsolete progressive call request (that was canceled, callee-leaved, etc) or a brand-new one — is to analyze the request id: is it the next one or previously used). 

I've looked into Nexus Router implementation — it doesn't keep track of the last used request ID, only checks for existence. 

I dug through the spec and can't find a description of what should happen if one of the peers (mostly router) receives a non-incremental request ID like previously used or +N (N>1) to current. Am I missing it or we indeed didn't describe it? 

So the questions are:

* What the peer should do when it receives a non-incremental request ID (in session scope)?
*  How do current router implementations work? Do they abort the connection? Nexus — doesn't care. (I don't say it is good or bad, just fact)
* Depending on this we can think of how to deal with the original question.

**Comment by ecorm** [comment]
> KSDaemon: The case when a router receives a progressive call with a request ID that doesn't exists in the router runtime can be treated as a starting new progressive call. The only way to distinguish whether is it an obsolete progressive call request (that was canceled, callee-leaved, etc) or a brand-new one — is to analyze the request id: is it the next one or previously used).

That's exactly how I do it in my work-in-progress router implementation.

> KSDaemon: I've looked into Nexus Router implementation — it doesn't keep track of the last used request ID, only checks for existence.

In my dealer implementation, I ended up having to keep track of the request ID of the last created call operation. A request ID above that value is considered a new call operation.

> KSDaemon: I dug through the spec and can't find a description of what should happen if one of the peers (mostly router) receives a non-incremental request ID like previously used or +N (N>1) to current. Am I missing it or we indeed didn't describe it?

Under [IDs](https://wamp-proto.org/wamp_latest_ietf.html#name-ids):

> IDs in the _session scope_ MUST be incremented by 1 beginning with 1 (for each direction - _Client-to-Router and Router-to-Client_) {#session_scope_id}

And under [Protocol Errors](https://wamp-proto.org/wamp_latest_ietf.html#name-protocol-errors) (recently added in PR #437):

> Receiving message with non-sequential session scope request ID, such as SUBSCRIBE, UNSUBSCRIBE, PUBLISH, REGISTER, UNREGISTER, CALL and YIELD.

And further below:

> In all such cases WAMP implementations:
> 1. MUST send an ABORT message to the offending peer, having reason wamp.error.protocol_violation and optional attributes in ABORT.Details such as a human readable error message.
> 2. MUST abort the WAMP session by disposing any allocated subscriptions/registrations for that particular client and without waiting for or processing any messages subsequently received from the peer,
> 3. SHOULD also drop the WAMP connection at transport level (recommended to prevent denial of service attacks)

> KSDaemon: How do current router implementations work? Do they abort the connection? Nexus — doesn't care. (I don't say it is good or bad, just fact)

In mine I abort with a protocol violation URI as per that recent PR.

**Comment by oberstet** [comment]
don't call me cynical, but it is fascinating that we now finally reached a design/development phase in WAMP where we need to dig down details of what should happen when you _combine multiple complex features_.

tbh, I'm not entirely sure if we can do that without some more formalization. but we can give it a try.

eg what are your routers doing with **complex features** like e.g.

[11.1. ](https://wamp-proto.org/wamp_latest_ietf.html#section-11.1)[Progressive Call Results](https://wamp-proto.org/wamp_latest_ietf.html#name-progressive-call-results)

[11.2. ](https://wamp-proto.org/wamp_latest_ietf.html#section-11.2)[Progressive Calls](https://wamp-proto.org/wamp_latest_ietf.html#name-progressive-calls)

[11.3. ](https://wamp-proto.org/wamp_latest_ietf.html#section-11.3)[Call Timeouts](https://wamp-proto.org/wamp_latest_ietf.html#name-call-timeouts)

[11.4. ](https://wamp-proto.org/wamp_latest_ietf.html#section-11.4)[Call Canceling](https://wamp-proto.org/wamp_latest_ietf.html#name-call-canceling)

[11.5. ](https://wamp-proto.org/wamp_latest_ietf.html#section-11.5)[Call Re-Routing](https://wamp-proto.org/wamp_latest_ietf.html#name-call-re-routing)

----

for example, what happens in a call with progressive results and `invocation == "roundrobin"?

what happens if such a call times out?

what happens if I cancel such an ongoing call?

now, what if the call not only produces progressive results, but _also_ (at the same time) receives progressive call arguments.

----

for ultimate challenge, consider when you mix into above the following (which Crossbar.io provides .. not sure we have spec text or even discussed that widely):

a callee (including a re-routed one) can be restricted when registering using `concurrency = N` in the maximum number of outstanding invocations the router will feed that endpoint with.

**XX.X Callee concurrency**

> this feature can be useful in two situation: in a multi-callee, scale-out situation you want to still limit the load to < number of cores per callee. the other situation: if you have a callee that can only work non-concurrently, strictly request-response, serialized (eg think of a physical machine that can only do 1 thing at a time and needs to finish before starting next job), use `concurrency == 1`

if the maximum is reached, the call can either be error'ed, or - if no invocation slot is available on any of the potentially multiple endpoints registered callees - the invocation can be internally queued, and subsequently (when slots become available) be processed

**XX.X Invocation queuing**

the maximum queuing again can be limited, and only once _that_ max (total) is reached, the call will be failed.

----

I lied;) thinking about this stuff, the real maximum complexity will be reached once you recognize that WAMP allows to dynamically register/unregister endpoints while a call directed to such an target is ongoing.

**Comment by ecorm** [comment]
> don't call me sarcastic?, but it is fascinating that we now finally reached a design/development phase in WAMP where we need to dig down details of what should happen when you combine multiple complex features.

You created a monster! :grin: 

Perhaps separate messages/mechanisms for streaming would have been better than trying to overload the RPC mechanism. I was not involved in the design of progressive results/calls; I'm just trying to implement them in my router and can't help but stumble into these issues that are not covered by the spec. I could just decide unilaterally what to do in my router when things are under-specified, but I prefer to raise awareness about them instead.

**Comment by ecorm** [comment]
Submitted PR #449

**Comment by Jopie64** [comment]
Just to bluntly throwing in my opinion. [Here](https://github.com/wamp-proto/wamp-proto/pull/445#issuecomment-1439377238) @ecorm said:

> One thing that's missing is what happens when a CALL message uses a request ID for a progressive call that no longer exists (due to the callee unregistering or leaving) and the caller has not yet received the wamp.error.cancelled error.

Which implies a caller cannot send more progressive call arguments for an outstanding invocation when a procedure is unregistered.

I think this is inconsistent with the decision that a callee _can_ finish an outstanding invocation with progressive results.
My opinion is that a caller should be able to send progressive arguments until an invocation is finished by a cancel from the caller or a result or error from the callee. An intermediate unregister should only mean the callee does not accept _new_ invocations but should not have any effect on current invocations.

**Comment by ecorm** [comment]
@Jopie64 When I wrote that, the apparent consensus on this proposal concerning UNREGISTER during an active call slipped my mind: https://github.com/wamp-proto/wamp-proto/issues/283#issuecomment-429542748

There still remains the use case where the callee leaves. There needs to be way to notify the caller so that it stops spamming progressive calls within that expired request. If the callee leaves, then the procedure no longer exists, so it seems we could use `wamp.error.no_such_procedure` for that notification.

The race condition between a progressive caller receiving `wamp.error.no_such_procedure` (due to callee leaving) and sending a subsequent progressive call is still a consideration, so my points concerning progressive `CALL`s referencing expired request IDs remain.

Thank you for your feedback!

**Comment by Jopie64** [comment]
Also I don't understand this...

> for example, what happens in a call with progressive results and invocation == "roundrobin"?
>
> what happens if such a call times out?
>
> what happens if I cancel such an ongoing call?
>
> now, what if the call not only produces progressive results, but also (at the same time) receives progressive call arguments.

Maybe I misunderstand something, but what is the discrepancy here?
As far as I understood the "roundrobin" option only influences an (initial) `INVOCATION`, and doesn't have any effect on progressive calls or call results of an ongoing call. Once a callee is selected by the chosen algorithm, it sticks to that callee as long as the call takes.

This is in our experience what the Crossbar router already does since we use both the round robin and progressive results feature...

**Comment by Jopie64** [comment]
@ecorm 
> There still remains the use case where the callee leaves. There needs to be way to notify the caller so that it stops spamming progressive calls within that expired request.

True.
Note however that the router already should send an `ERROR` respons with aborted I believe when an ongoing call is aborted due to the callee leaving. So an extra response due to an undeliverable progressive argument is then only an (I think good) addition.

**Comment by ecorm** [comment]
For my race condition scenario, I think there's no need to invent a new `wamp.error.no_such_call` error URI. The dealer only needs to resend `wamp.error.no_such_procedure` if the caller missed the previous one before sending a subsequent progressive call.

The caller would then have to ignore the second `wamp.error.no_such_procedure` ERROR message it receives.

This would be easier to explain with a sequence diagram. I'll whip one up.

I never took the time to study Shared Registrations, so I don't know how they fit into all this.

**Comment by oberstet** [comment]
> Maybe I misunderstand something, but what is the discrepancy here?

for this particular case (invocation policy + progressive results), what you describe is what Crossbar.io does! 

> As far as I understood the "roundrobin" option only influences an (initial) INVOCATION

yes, for a CALL that receives progressive results, but is not invoking with progressive call arguments, yes, that's how it works:

https://github.com/crossbario/crossbar/blob/427cf58e42014499a623d4c17e06c6e3440273f9/crossbar/router/dealer.py#L847

> Once a callee is selected by the chosen algorithm, it sticks to that callee as long as the call takes.

again, yes for the specific combination, the call will be kept open by the router and further YIELD'ed results forwarded to the caller

https://github.com/crossbario/crossbar/blob/427cf58e42014499a623d4c17e06c6e3440273f9/crossbar/router/dealer.py#L896

> it sticks to that callee as long as the call takes.

yes:

https://github.com/crossbario/crossbar/blob/427cf58e42014499a623d4c17e06c6e3440273f9/crossbar/router/dealer.py#L1349

-----

above Crossbar.io behavior IMO makes sense - it is least surprising for users and is useful

as far as I see, even this behavior (assuming other implementors would agree) isn't described in the spec. it is "implied by logic" at best;)

the behavior when you now mix in:

* cancel (caller or callee side)
* timeout
* concurrent register/unregister

is also not covered (not in the spec, and tbh, I forgot if or how we have code in crossbar to deal with that)

**Comment by oberstet** [comment]
> Note however that the router already should send an `ERROR` response ...

surprisingly, it seems we have even covered some races, like a call with progressive results receiving an error after being canceled

https://github.com/crossbario/crossbar/blob/427cf58e42014499a623d4c17e06c6e3440273f9/crossbar/router/dealer.py#L1474

thing is: I wouldn't bet my life on that we've covered _all_ cases.

as @ecorm said: we've created a monster;)

I am happy nevertheless, because a) the monster arises from combining cleanly separated and orthogonal concepts. and b): it can be fixed, as the protocol design is sound, and "only" needs formalization, and then a provable correct router;)

**Comment by ecorm** [comment]
This is my first time embedding a Mermaid diagram in GitHub markdown, so please go easy on me. :grin:

While drawing the diagram, it occurred to me that the simplest solution might be for the Dealer to keep the progressive call request alive until the next progressive CALL. Only then does it return a `wamp.error.no_such_procedure`. This avoids the caller receiving duplicate ERROR messages due to race conditions.

```mermaid
sequenceDiagram
    participant Caller
    participant Dealer
    participant Callee
    participant Request

    Caller->>Dealer: CALL (progress)
    Dealer->>Request: <<create>>
    Dealer->>Callee: INVOCATION (progress)

    Callee->>Dealer: YIELD
    Dealer->>Caller: RESULT


    Callee->>Dealer: LEAVE
    Caller->>Dealer: CALL (progress)
    Dealer->>Request: <<destroy>>
    Dealer->>Caller: ERROR (no_such_procedure)
```

**Comment by oberstet** [comment]
I like that!

one can immediately see: when a router can process a `CALL (progress)` while `LEAVE + <<destroy>>` is not atomic, this is a race.

but why isn't there an ERROR returned to the Caller when the Callee left - and thus before an ERROR (no_such_proc) would later be returned when a CALL (progress) is racing in?

**Comment by oberstet** [comment]
essentially: when the Callee leaves while a progressive call is on-going, there is a race for

* ERROR (some_callee_error) - newly incoming CALL is answered after an ERROR has been already sent for the disappeared Callee
* ERROR (no_such_procedure) - newly incoming CALL is answered after Callee is gone but before an ERROR for that was sent already

**Comment by Jopie64** [comment]
@oberstet 
> above Crossbar.io behavior IMO makes sense - it is least surprising for users and is useful
>
> as far as I see, even this behavior (assuming other implementors would agree) isn't described in the spec. it is "implied by logic" at best;)

Ah. So the point was that even when the behaviour is logical, it wasn't/isn't explicitly described in the spec. So indeed my misunderstanding 😊

**Comment by ecorm** [comment]
> but why isn't there an ERROR returned to the Caller when the Callee left - and thus before an ERROR (no_such_proc) would later be returned when a CALL (progress) is racing in?

Did you mean this? Note that Mermaid does not support delayed messages (with the non-horizontal lines), so I had to emulate them with a "Start sending ERROR" and "Received ERROR" pair.

```mermaid
sequenceDiagram
    participant Caller
    participant Dealer
    participant Callee
    participant Request

    Caller->>Dealer: CALL (progress)
    Dealer->>Request: <<create>>
    Dealer->>Callee: INVOCATION (progress)

    Callee->>Dealer: YIELD
    Dealer->>Caller: RESULT


    Callee->>Dealer: LEAVE
    Dealer->>Request: <<destroy>>
    Dealer->>Caller: Start sending ERROR (cancelled)
    Caller->>Dealer: CALL (progress)
    Dealer->>Caller: Received ERROR (cancelled)
    Dealer->>Caller: ERROR (no_such_procedure)
```

**Comment by Jopie64** [comment]
Nice that mermaid also works in GitHub 👍

@ecorm 
For me it's unclear in the diagram whether the `YIELD` / `RESULT` is progressive or not. When not explicitly mentioned it defaults to not, but that would mean the call ends and that progressive arguments should already be denied even without the leave.

**Comment by ecorm** [comment]
> For me it's unclear in the diagram whether the YIELD / RESULT is progressive or not. When not explicitly mentioned it defaults to not, but that would mean the call ends and that progressive arguments should already be denied even without the leave.

Ah yes, thanks for pointing that out. I'll edit my sequence diagrams above to indicate the YIELD / RESULT is progressive. I don't even need YIELD / RESULT in my diagrams if we limit ourselves to Progressive Calls only without Progressive Call Results (this stuff gets confusing too easily).

**Comment by ecorm** [comment]
> or put yet another way: in your router, the actions
>
>    receive LEAVE from callee
>    expire CALL request object in dealer
>    send ERROR to caller (this is currently missing in the diagram)

Not sending an ERROR immediately upon the Callee leaving was intentional to avoid the scenario where two ERRORs are sent to the Callee.

>    destroy CALL request object in dealer
>
> are not executed as 1 transaction, right?

Marking the Request as expired is part of the LEAVE transaction. Destroying the Request and returning an ERROR to the Caller is a different transaction. I added a note in the diagram that indicates the Dealer waits for the next CALL before destroying the Request.

**Comment by ecorm** [comment]
> are not executed as 1 transaction, right?

I edited both diagrams to add activation rectangles to make it clearer which operations are included within transactions.

**Comment by oberstet** [comment]
> Did you mean this?

yes, this already captures the essence much better (in my eyes).

if we want to make such diagrams into a semi-formal language, it needs even more precision IMO.

1. the 2 arrows `<<destroy>>` and `ERROR (canceled)` should originate at the same origin (time within dealer)
2. the whole column `Request` (I get it -- but ..) is "alien", as it is not a WAMP peer and arrows are not WAMP messages - but a dealer internal artifact and dealer internal function calls (I guess)

and 3.

> Note that Mermaid does not support delayed messages (with the non-horizontal lines), so I had to emulate them using parallel blocks.

if we want to model race conditions in distributed systems like WAMP, we must model "the network" in between

e.g. if you'd introduced a column "Network", then you can model races between _sending_ `CALL (progress)` and _receiving_ `ERROR (cancelled)` properly (which currently isn't the case in the diagram, as the Caller send another `CALL (progress)` _after_ it received `ERROR (cancelled)` - and that is a protocol error)

**Comment by Jopie64** [comment]
> I added a note in the diagram that indicates the Dealer waits for the next CALL before destroying the Request.

Maybe misunderstanding it again, but why would you do that? After a `LEAVE` the request end should already have been communicated to the caller right? Why wait for another `CALL`? What if it never comes?

**Comment by oberstet** [comment]
@Jopie64 I think the misunderstanding is not a misunderstanding, but comes from the problems of capturing semantics of race conditions in distributed systems using message sequence charts where there the *Network* isn't an actor itself. the vertical grey boxes used in the diagram to signal "run time > 0" for some step capture something happening _within_ a peer - and this is a source of races, but only one. the other source of races is the fact that "Network" isn't latency == 0. If we had another column, that could again be mapped to grey boxes. Eg columns

1. Caller
2. Network
3. Dealer
4. Network
5. Callee

**Comment by ecorm** [comment]
> 1. the 2 arrows <<destroy>> and ERROR (canceled) should originate at the same origin (time within dealer)

I added activation rectangles that make it clearer which operations are executed "atomically".

>  2. the whole column Request (I get it -- but ..) is "alien", as it is not a WAMP peer and arrows are not WAMP messages - but a > dealer internal artifact and dealer internal function calls (I guess)

Request is indeed an object that is internal to the Dealer. Perhaps it's not formally correct UML, but my goal here is to achieve understanding and consensus on the intended behavior, not formalism. We can design a more formal sequence diagram for the spec later.

**Comment by ecorm** [comment]
> Maybe misunderstanding it again, but why would you do that? After a LEAVE the request end should already have been communicated to the caller right?

I did that to avoid the scenario where two ERROR messages are sent back to the Caller.

> Why wait for another CALL? What if it never comes?

We're already facing this possibility of a subsequent progressive CALL never coming in the "normal" case where the Callee doesn't leave.

**Comment by oberstet** [comment]
fwiw, what I mean is sth like this (got tired .. the right grey boxes and all the error stuff is missing, but should be good enough to understand)

```mermaid
sequenceDiagram
    participant Caller
    participant Transport1
    participant Dealer
    participant Transport2
    participant Callee

    Caller->>Transport1: send CALL (progress)
    activate Transport1
        Transport1->>Dealer: receive CALL (progress)
    deactivate Transport1

    Dealer->>Transport2: send INVOCATION (progress)
    activate Transport2
        Transport2->>Callee: receive INVOCATION (progress)
    deactivate Transport2

    Callee->>Transport2: send YIELD (progress)
    activate Transport2
        Transport2->>Dealer: receive YIELD (progress)
    deactivate Transport2
 
    Dealer->>Transport1: send RESULT (progress)
    activate Transport1
        Transport1->>Caller: receive RESULT (progress)
    deactivate Transport1
```

**Comment by ecorm** [comment]
To make it clearer, I edited the diagrams so that an open arrow (asynchronous message) is used for network messages and a solid arrow is used for function calls to the internal Request object. This is one thing I dislike about UML: stylistic things like open/solid arrows, or _italics_ conveys semantics.

> fwiw, what I mean is sth like this (got tired .. the right grey boxes and all the error stuff is missing, but should be good enough to understand)

I understand what you mean, but I'm afraid mixing in the transport layer will just make the diagrams too busy and difficult to understand. They also would be difficult to fit across 80 columns in RFC plain text format.

I really like the angled message lines that convey delay semantics, but I don't know if those can be easily rendered in plain text.

**Comment by KSDaemon** [comment]
A community of sequence diagrammers :)

This is mine versions. Here is the first one that (I hope) reflects the @ecorm first version with 2 errors:

```mermaid
sequenceDiagram
    participant Caller
    participant NetRD as Network
    participant Dealer
    participant NetDR as Network
    participant Callee

    Caller->>Dealer: CALL (start progress)
    Dealer-->>Dealer: <create request>
    Dealer->>Callee: INVOCATION
    Caller->>NetRD: CALL (inprogress 1)
    activate NetRD
    Callee->>Dealer: GOODBYE
    Dealer->>Callee: GOODBYE
    Dealer->>Caller: CALL ERROR
    Dealer-->>Dealer: <destroy request>
    NetRD->>Dealer: CALL (inprogress 1)
    deactivate NetRD
    Dealer->>Caller: CALL ERROR (no_such_procedure)
```

If I understand correctly this is how double errors can be avoided:

```mermaid
sequenceDiagram
    participant Caller
    participant NetRD as Network
    participant Dealer
    participant NetDR as Network
    participant Callee

    Caller->>Dealer: CALL (start progress)
    Dealer-->>Dealer: <create request>
    Dealer->>Callee: INVOCATION
    Caller->>NetRD: CALL (inprogress 1)
    activate NetRD
    Callee->>Dealer: GOODBYE
    Dealer->>Callee: GOODBYE
    Dealer->>Caller: CALL ERROR
    NetRD->>Dealer: CALL (inprogress 1)
    deactivate NetRD
    Note right of Dealer: as request is not destroyed<br>dealer knows that it already<br>sent an error
    Dealer-->>Dealer: do nothing
    
```

**Comment by oberstet** [comment]
@KSDaemon that last diagram shows what I meant: pnly the presence of and the fact that "Network" takes latency>0 allows to show the race in the 2nd CALL sending but not yet receiving at the Dealer when the CALL ERROR was already sent by the Dealer.

even with infinite fast Caller, Dealer and Callee, _those_ Network related races are still present.

only if you assume both infinite fast Peers _and_ infinite fast Network the races disappear

**Comment by KSDaemon** [comment]
yep! 
I made transparent network arrows because they are out of interest to focus on `call inprogress` latency

**Comment by ecorm** [comment]
> Here is the first one that (I hope) reflects the ecorm first version with 2 errors:

@KSDaemon That would be equivalent to my **second** version which has the two errors.

> If I understand correctly this is how double errors can be avoided:

That would be a variant of my first version, where in yours the `ERROR` is sent right after the Callee leaves. In yours a `wamp.error.cancelled` URI would make sense. I'm not entirely sure the race condition is resolved, though; I need to think about it some more.

> Me: I understand what you mean, but I'm afraid mixing in the transport layer will just make the diagrams too busy and difficult to understand. They also would be difficult to fit across 80 columns in RFC plain text format.

Seeing how KSDaemon drew it, I admit it doesn't look as busy as I thought it would. But can it fit in RFC plain text format?

**Comment by KSDaemon** [comment]
Let's think about it more!

> But can it fit in RFC plain text format?

I think `yes`. Just look at our [Sessions](https://wamp-proto.org/wamp_latest_ietf.html#name-sessions) for example :)

**Comment by KSDaemon** [comment]
> Why wait for another CALL? What if it never comes?

> We're already facing this possibility of a subsequent progressive CALL never coming in the "normal" case where the Callee doesn't leave.

Я вот тут подумал: что для защиты от ДоС роутер может предусматривать максимальное количество вызовов, которое может одновременно сделать пир. Как думаете: стоит ли это описывать в спеке или оставить на откуп имплементациям? Но хорошо бы как минимум формализовать URI ошибки.

I was thinking: to protect against DoS (kind of starting progressive calls and not finishing them), the router may specify the maximum number of calls that the peer can make simultaneously. What do you think: should this be described in the spec or left to implementation? But it would be good to at least formalize URI errors.

**Comment by ecorm** [comment]
I think it's impossible to avoid a second `ERROR` in either KSDaemon's or my variation where we keep the internal Request object alive. All we're doing us pushing the race condition to the next CALL that hasn't been received yet. I'm borrowing KSDaemon's second diagram to illustrate this:

```mermaid
sequenceDiagram
    participant Caller
    participant NetRD as Network
    participant Dealer
    participant NetDR as Network
    participant Callee

    Caller->>Dealer: CALL (start progress)
    Dealer-->>Dealer: <create request>
    Dealer->>Callee: INVOCATION
    Caller->>NetRD: CALL (inprogress 1)
    activate NetRD
    Callee->>Dealer: GOODBYE
    Dealer->>Callee: GOODBYE
    NetRD->>Dealer: CALL (inprogress 1)
    deactivate NetRD
    Caller->>NetRD: CALL (inprogress 2)
    activate NetRD
    Note right of Dealer: as request is not destroyed<br>dealer sends an error<br>instead of no_such_procedure
    Dealer->>Caller: CALL ERROR
    Dealer-->>Dealer: <delete request>
    NetRD->>Dealer: CALL (inprogress 2)
    deactivate NetRD
    Note right of Dealer: as request is destroyed<br>dealer sends no_such_procedure
    Dealer->>Caller: CALL ERROR
```

**Comment by KSDaemon** [comment]
Yeah! You're right! The only possible way of not sending 2nd error - is to keep the request object forever. But that's almost certainly not acceptable.

So i think we should make in straight forward: Send ERROR, if next call come in place - send `no_such_procedure`.

**Comment by Jopie64** [comment]
Question in between:
Is it worth the trouble solving the second error problem?

I mean, after the first error, a client should already cleanup the call logic so the second error doesn't match an ongoing call any longer so the client could simply ignore it.

In the WAMP spec we could simply state that error responses can be received for calls that do not exist (anymore) due to races and that clients should ignore them. Just like the dealer can receive `YIELD` for invocations that no longer exist due to races, that should then be ignored.

**Comment by KSDaemon** [comment]
The caller may process that gracefully on its own side: as it has info about what the call error relates to - it can treat it not like a real error.

**Comment by oberstet** [comment]
for the 2nd CALL, there are 2 variants:

1. `CALL (inprogress 2 with request ID from initial call)`
2. `CALL (2nd new one with new request ID)`

I would argue "no such procedure" should only be returned for case 2. but actually, *why not simply ignore the 2nd one and send back nothing (the request ID is for an old, completed/errored call)*?

for case 1, it should be "no such procedure" (assuming there aren't other shared registrations for that procedure)

> in any case: I like this diagram format a lot. it allows precise communication. also the reflexive arrows for peer-internal stuff like "create request"

**Comment by KSDaemon** [comment]
Is the next statement correct?

If the dealer receives a progressive CALL with a Request ID that is `<=` the latest Request ID from the client session and there is no ongoing invocation info in the router runtime - that means that the router already sent an error/final result to the CALLER and may do nothing.

**Comment by KSDaemon** [comment]
Okay) Super productive brainstorming folks! Thank you all! But I need to sleep! (at least a bit)

**Comment by oberstet** [comment]
> Okay) Super productive brainstorming folks! Thank you all! But I need to sleep! (at least a bit)

yep!! very productive. but same here: sleeeep;)

**Comment by ecorm** [comment]
> for case 1: why not simply ignore the 2nd CALL (with old request ID) and send back nothing (the request ID is for an old, completed/errored call)?

Because of the race caused by the Callee leaving, it is outside the control of the Caller: the Caller cannot be blamed. If there is never any ERROR response to that CALL, the Caller ends up with that CALL stuck in its pending request container.

By time the Dealer receives that "obsolete" CALL, it may no longer remember anything about that old progressive call request. The callee is gone and so is the registration for that procedure (assuming no shared registrations). With the current information available to it, what other error could the Dealer return other than `wamp.error.no_such_procedure`?

I'm going take the time to read up on Shared Registrations to be in a better position to contribute towards a solution for this issue that I raised.

**Comment by ecorm** [comment]
Now that we've established that there's no utility in keeping the Request alive after a Callee leaves, we're left with the behavior shown in KSDaemon's first diagram (and my second diagram). The question remains what URI should be used to report to the Caller that the Callee has left in the middle of a ongoing progressive call?

We could use `wamp.error.cancelled`, but that might by confusing to the Caller because they never sent a `CANCEL` message. If we want to avoid that confusion, we have to come up with a new error URI. Some ideas:

- `wamp.error.expired`
- `wamp.error.call_expired`
- `wamp.error.progressive_call_expired`
- `wamp.error.callee_left`
- `wamp.error.call_is_kaput`

**Comment by Jopie64** [comment]
> The question remains what URI should be used to report to the Caller that the Callee has left in the middle of an ongoing progressive call?

Why would that not be the same as with a non-progressive call?

**Comment by ecorm** [comment]
> Why would that not be the same as with a non-progressive call?

Good point. But is it mandated anywhere what the error URI should be for non-progressive calls?

**Comment by Jopie64** [comment]
> Good point. But is it mandated anywhere what the error URI should be for non-progressive calls?

Also good point 😊
But I think it's better to specify that instead of specifying it solely for progressive calls.
Maybe take inspiration from Crossbar router? 😊

**Comment by ecorm** [comment]
> I'm going take the time to read up on Shared Registrations to be in a better position to contribute towards a solution for this issue that I raised.

I read up on Shared Registrations, and it seems that it's only applicable to this issue when the last available registered Callee has left. The logic should then be the same as in the single registration case.

**Comment by KSDaemon** [comment]
My thoughts on URI list:

* wamp.error.expired
* wamp.error.call_expired
* wamp.error.progressive_call_expired
* wamp.error.callee_left
* wamp.error.call_is_kaput

I don't like `*.expired` variants as they do not really reflect the essence of the problem. Call is not expired, just the Callee was blown away :)

I like variant `wamp.error.callee_left`. Do we have other cases when this not reflect the case? I mean do we have cases when we can not proceed the [ongoing] call while caller is alive due to some other reasons? Can't find one....

**Comment by ecorm** [comment]
> Why would that not be the same as with a non-progressive call?

It turns out this issue is not applicable to non-progressive calls. Upon a non-progressive CALL, the Dealer knows if there is a matching registration or not. If the Callee already send a GOODBYE message and cleared its internal registry of RPC handlers, it will respond to the INVOCATION with an ERROR which will be propagated back to the CALLER.

Let me illustrate:

```mermaid
sequenceDiagram
    participant Caller
    participant Dealer
    participant Callee

    Caller->>Dealer: CALL (progress=false)
    activate Dealer
    Dealer-->>Dealer: <create request>
    Callee->>Dealer: GOODBYE (queued)
    Dealer->>Callee: INVOCATION
    deactivate Dealer
    Callee->>Dealer: INVOCATION ERROR (queued)
    Dealer->>Callee: GOODBYE
    Dealer->>Caller: CALL ERROR
```

**Comment by ecorm** [comment]
> I mean do we have cases when we can not proceed the [ongoing] call while caller is alive due to some other reasons? Can't find one....

There is such a case with the [Call Re-Routing](https://wamp-proto.org/wamp_latest_ietf.html#name-call-re-routing) feature.

We could reuse the  `wamp.error.no_available_callee` error from the `single` invocation policy of that feature.

**Comment by KSDaemon** [comment]
@Jopie64 in case of a `non-progressive` call Caller makes only one call request and then is waiting for results (or error).

**Comment by KSDaemon** [comment]
> We could reuse the wamp.error.no_available_callee error from the single invocation policy of that feature.

Actually not a bad idea. In general, I would prefer not to create a new URI for every case and try to generalize and reuse existing ones.

**Comment by oberstet** [comment]
> In general, I would prefer not to create a new URI for every case and try to generalize and reuse existing ones.

yes, in general. however, the counter balancing reason (for "more URIs") is when the peer can/wants to react differently depending on the different situation signaled by the different error URI

eg in case of `wamp.error.no_available_callee`, there is no point in retrying (immediately). the service the client wants to use is gone (for now). what a client could do: register via the WAMP meta API to notice once a new (first) callee becomes online again

**Comment by ecorm** [comment]
> eg in case of `wamp.error.no_available_callee`, there is no point in retrying (immediately). the service the client wants to use is gone (for now). what a client could do: register via the WAMP meta API to notice once a new (first) callee becomes online again

Sorry, it's still not clear to me if reusing `wamp.error.no_available_callee` is acceptable to you, or if you prefer a separate URI such as `wamp.error.callee_left` (or something else).

**Comment by KSDaemon** [comment]
> yes, in general. however, the counter balancing reason (for "more URIs") is when the peer can/wants to react differently depending on the different situation signaled by the different error URI

Well, sure! You always have to use your common sense.

In this concrete case I think reusing `wamp.error.no_available_callee` suits well.

**Comment by Jopie64** [comment]
@ecorm @KSDaemon 
I love that mermaid stuff! Make things clearer.

I don't mean the 'race-case'. What I mean is the simple case that a call is cancelled because of a callee leave like this:

```mermaid
sequenceDiagram
    participant Caller
    participant Dealer
    participant Callee

    Caller->>Dealer: CALL (progress=false)
    Dealer->>Callee: INVOCATION
    Callee->>Dealer: GOODBYE
    Dealer->>Caller: CALL ERROR (cancelled)
```

My idea was this could be the same for progressive calls / call responses like this:

```mermaid
sequenceDiagram
    participant Caller
    participant Dealer
    participant Callee

    Caller->>Dealer: CALL (progress=true)
    Dealer->>Callee: INVOCATION
    Callee->>Dealer: YIELD (progress=true)
    Dealer->>Caller: RESULT (progress=true)
    Caller->>Dealer: CALL (still the same call instance)
    Dealer->>Callee: INVOCATION
    Callee->>Dealer: GOODBYE
    Dealer->>Caller: CALL ERROR (cancelled)
```

Not sure whether that was already cleared out in the specs, but I may have a hidden assumption that progressive calls are 'sticky' to a single callee even with shared registrations. So once the CALL is established and a callee is selected via shared distribution algorithm, progressive calls are sent to the same callee. So that means, when the selected callee leaves, the call is cancelled even when there're other shared callees being able to handle the same call.

Anyway with that in mind, I don't see why there should be a difference with progressive and non-progressive calls, even in case of races... At most the 'problem' that the dealer will send 2 cancel responses in cases like this:

```mermaid
sequenceDiagram
    participant Caller
    participant Network
    participant Dealer
    participant Callee

    Caller->>+Network: CALL (progress=true)
    Network->>-Dealer: CALL (progress=true)
    Dealer->>Callee: INVOCATION
    Callee->>Dealer: GOODBYE
    Caller->>+Network: progressive CALL
    Dealer->>+Network: CALL ERROR (cancelled)
    Network->>-Caller: CALL ERROR (cancelled)
    Network->>-Dealer: progressive CALL
    note right of Dealer: Call not active in dealer<br>but old sequence nr<br>So race could be detected here<br>Still it could send error anyway...
    Dealer->>+Network: CALL ERROR (cancelled)
    Network->>-Caller: CALL ERROR (cancelled)
```

Sorry for the wall of text. I want it to be clear :)

**Comment by ecorm** [comment]
@Jopie64 Ok, I see it now. Thanks for clarifying. Yes, I agree with you that that behavior should be defined for the non-progressive case and it would also apply for the progressive case.

I had also originally suggested `wamperror.cancelled` as the error URI, but changed my mind because it would be confusing to a Caller who might have also sent a CANCEL message. See our discussion above about whether to reuse `wamp.error.no_available_callee` (from the Call Rerouting feature), or invent a new one like `wamp.error.callee_left`.

> Not sure whether that was already cleared out in the specs, but I may have a hidden assumption that progressive calls are 'sticky' to a single callee even with shared registrations.

Stickness was discussed in #446 and the discussion then moved to PR #449.

**Comment by oberstet** [comment]
> Sorry, it's still not clear to me if reusing wamp.error.no_available_callee is acceptable to you, or if you prefer a separate URI such as wamp.error.callee_left (or something else).

I agree with @KSDaemon on this one: +1 for reuse of `wamp.error.no_available_callee`

> Not sure whether that was already cleared out in the specs, but I may have a hidden assumption that progressive calls are 'sticky' to a single callee even with shared registrations. So once the CALL is established and a callee is selected via shared distribution algorithm, progressive calls are sent to the same callee. So that means, when the selected callee leaves, the call is cancelled even when there're other shared callees being able to handle the same call.

That's also my understanding (it makes sense), but I also lost sight whether we have it already (coming) in spec

**Comment by KSDaemon** [comment]
> Not sure whether that was already cleared out in the specs, but I may have a hidden assumption that progressive calls are 'sticky' to a single callee even with shared registrations. So once the CALL is established and a callee is selected via shared distribution algorithm, progressive calls are sent to the same callee. So that means, when the selected callee leaves, the call is cancelled even when there're other shared callees being able to handle the same call.

Regarding shared and progressiveness - there is a CALL option `sticky`which controls how each progressive call is routed. See the spec here: https://wamp-proto.org/wamp_latest_ietf.html#section-11.2-25

**Comment by oberstet** [comment]
> Regarding shared and progressiveness - there is a CALL option stickywhich controls how each progressive call is routed. 

ohhh. you are right. `sticky == false` with `progress == true` -- and then a failing / timed-out / etc callee. that's the tricky one.

rgd terminology: with a callee doing a shared registration, how do we call each individual instance of the registered procedure?

I am using "endpoint", as in:

one *procedure* can have zero, one or multiple *endpoints* registered. in one *session* or in multiple sessions. and a *client* (a process) can have zero, one (usual case) or even multiple sessions

**Comment by ecorm** [comment]
> Regarding shared and progressiveness - there is a CALL option stickywhich controls how each progressive call is routed. See the spec here: https://wamp-proto.org/wamp_latest_ietf.html#section-11.2-25

It's not (yet) specified whether the `sticky` option is sticky (as I defined "sticky" in my recent PR). That is, can a subsequent progressive call alter the `sticky` property of the request?

I should find a synonym for "stickiness" in my PR, because it will be confused with this `sticky` option I was previously unaware of.

**Comment by oberstet** [comment]
> I should find a synonym for "stickiness" in my PR, because it will be confused with this `sticky` option I was previously unaware of.

yes, I also only now realize. those are different beasts.

1. "sticky" as in the context of shared registrations (this is about forwarding progressive arguments or results)
2. "sticky" as in the call details parameters are chosen via the very first initial message, and all subsequent progressive messages must follow

**Comment by KSDaemon** [comment]
:))) OMG)))

`endpoint` - is a good term for that imo.

> It's not (yet) specified whether the sticky option is sticky (as I defined "sticky" in my recent PR). That is, can a subsequent progressive call alter the sticky property of the request?

As we discussed above - this option should also be frozen with the value from the first call

**Comment by ecorm** [comment]
> As we discussed above - this option should also be **frozen** with the value from the first call

Ah, there's the synonym/metaphor I've been looking for: _frozen_. Thanks!

**Comment by Jopie64** [comment]
> Regarding shared and progressiveness - there is a CALL option sticky which controls how each progressive call is routed. See the spec here: https://wamp-proto.org/wamp_latest_ietf.html#section-11.2-25

Ah, thanks for pointing that out. I'm surprised to read that progressive calls can also _not_ be sticky. I feel that leads to all kinds of weird implications. E.g., what happens with call cancellation?

```mermaid
sequenceDiagram
    participant Caller
    participant Dealer
    participant Callee 1
    participant Callee 2

    Caller->>Dealer: CALL (progress=true)
    Dealer->>Callee 1: INVOCATION
    Caller->>Dealer: CALL (still the same call instance)
    Dealer->>Callee 2: INVOCATION
    Caller->>Dealer: CANCEL
    note right of Dealer: *Confused*
```

Dealer confused because, to which callee should it route cancellation?

* _None_<br>
  Could cause leak at callees

* _Callees involved by the dealer_<br>
  Feels random, but probably the best in this scenario.

* _Callees with shared registration of this procedure_<br>
  But what if one of the callees unregisters that procedure during the call? Should dealer consider that callee done with the call already?

---
Or another scenario: when one of the callees considers the call done, should the dealer communicate that to other involved callees?

```mermaid
sequenceDiagram
    participant Caller
    participant Dealer
    participant Callee 1
    participant Callee 2

    Caller->>Dealer: CALL (progress=true)
    Dealer->>Callee 1: INVOCATION
    Caller->>Dealer: CALL (still the same call instance)
    Dealer->>Callee 2: INVOCATION
    Callee 1->>Dealer: YIELD(progress=true)
    Dealer->>Caller: RESULT(progress=true)
    Callee 2->>Dealer: Final YIELD
    Dealer->>Caller: Final RESULT
    note right of Dealer: Shouldn't callee 1 know this?
    Callee 1->>Dealer: YIELD(progress=true)
    note right of Dealer: Now what?
```

And if so, what should it send? Cancel? And how should it react to progressive results from other callees that are unaware of the call being done? (Which can now also happen due to races.)

---
All this isn't clear to me reading the spec...
My blunt opinion would be to remove this non-sticky progressive call option entirely, cause I also can't think of a good usecase. But if there is a good usecase, and it's worth the effort to keep this in my eyes complex feature, at least specify what happens in these kinds of scenario's...

**Comment by ecorm** [comment]
> I am using "endpoint", as in:

"Endpoint" already has meanings in network communications.

From [Wikipedia](https://en.wikipedia.org/wiki/Communication_endpoint):

> A **communication endpoint** is a type of communication network node. It is an interface exposed by a communicating party or by a communication channel. An example of the latter type of a communication endpoint is a publish-subscribe topic [1] or a group in group communication systems.[2]

> rgd terminology: with a callee doing a shared registration, how do we call each individual instance of the registered procedure?

How about "candidate", "slot", or "enrollment"?

**Comment by oberstet** [comment]
> "Endpoint" already has meanings in network communications.

As does "Transport". IMO, we should aim for consistency and "least surprise" / suggestive terms - *within the specific context of WAMP*

> I'm surprised to read that progressive calls can also not be sticky.

yeah, we need to be precise/consistent IMO rgd terminology, e.g. one call with "progressive call arguments" and/or "progressive call results".

One "progressive call" is a call with at least one of "progressive call arguments" or "progressive call results" (or both).

And then there are calls which use "shared registrations" - that is the callee-side. There is no directly corresponding pendent "shared caller". A single call always originates from a single session.

**Comment by ecorm** [comment]
> All this isn't clear to me reading the spec...
> My blunt opinion would be to remove this non-sticky progressive call option entirely, cause I also can't think of a good usecase. > But if there is a good usecase, and it's worth the effort to keep this in my eyes complex feature, at least specify what happens in these kinds of scenario's...

One use case I can think of is parallel processing of real-time chunks of data. But an application can implement that with a single Callee serving as a mediator, which distributes chunks to other peers via RPC or pub/sub and then collects/accumulates the partial results.

**Comment by ecorm** [comment]
Perhaps the debate over the usefulness of the Progressive Calls `sticky` option should be continued in a separate issue? Or is it related to this issue where a Callee leaves in the middle of an RPC?

**Comment by oberstet** [comment]
> But an application can implement that 

that would be a design fail in my eyes, since it moves routing logic into clients. that is exactly what WAMP wants to achieve: a clean separation between routing and application code.

**Comment by oberstet** [comment]
yeah, maybe move .. or continue here. there is no escape, we need to drill down these deep holes .. here .. or elsewhere. anyways, need to catch some sleep, I'll check back tomorrow!

**Comment by ecorm** [comment]
> that would be a design fail in my eyes, since it moves routing logic into clients. that is exactly what WAMP wants to achieve: a clean separation between routing and application code.

Moving the routing logic to clients also admittedly involves more traffic to/from the router.

**Comment by ecorm** [comment]
> yeah, maybe move .. or continue here. there is no escape, we need to drill down these deep holes .. here .. or elsewhere. anyways, need to catch some sleep, I'll check back tomorrow!

Emergent Complexity Whac-A-Mole :grin:

**Comment by ecorm** [comment]
So, concerning a Callee leaving, I think all we have to do now is add something like this to the spec:

> When the last available *Callee* registered to an active procedure call leaves after a *Dealer* receives a `CALL`, but before that *Dealer* receives a `YIELD` or `ERROR` from the *Callee*, the *Dealer* shall send an `ERROR` message back to the *Caller* with URI `wamp.error.no_available_callee`.

The `wamp.error.no_available_callee` URI would have to be moved to the basic profile section.

Concerning request IDs, we'd also need to add something like this:

> When the *Progressive Calls* feature is supported, a *Caller* MAY use a prior request ID in a `CALL` message to reference an already active progressive call. However, if a `CALL` request ID refers to an active request that is not a progressive call, then that shall be considered a protocol violation.
>
> When checking `CALL` request IDs, if a request with the given ID is no longer active, then it shall NOT be considered a protocol violation. This is necessary to cope with the race condition of a *Caller* sending a subsequent progressive `CALL` while a *Dealer* concurrently sends an `ERROR` or final `RESULT` back to the *Caller*.

**Comment by Jopie64** [comment]
> When checking CALL request IDs, if a request with the given ID is no longer active, then it shall NOT be considered a protocol violation.

This is no matter what whether the call used to be progressive or not. Otherwise the dealer needs to remember old call instances. Should that be explicitly stated or does it already follow from the text?

**Comment by KSDaemon** [comment]
> My blunt opinion would be to remove this non-sticky progressive call option entirely, cause I also can't think of a good usecase. > But if there is a good usecase, and it's worth the effort to keep this in my eyes complex feature, at least specify what happens in these kinds of scenario's...

In addition to @ecorm case. Let's say you need to send your transport coordinates/parameters. As there are a huge number of moving cars/tracks you can optimize the backend and create a shared registration RPC for processing that data. Making calls progressive can be more optimal for all sides, but most important for the dealer: as it doesn't need to instantiate a new request object in runtime, no need to generate Ids, etc. Just reuse the existing one.

> But an application can implement that with a single Callee serving as a mediator, which distributes chunks to other peers via RPC or pub/sub and then collects/accumulates the partial results.

> that would be a design fail in my eyes, since it moves routing logic into clients. that is exactly what WAMP wants to achieve: a clean separation between routing and application code.

Exactly the same thoughts as @oberstet! That is what we aim for with WAMP: abstraction and loose coupling between peers :)

@ecorm - great additions to the spec. I think it clarifies the things we discussed so heart!

> When checking CALL request IDs, if a request with the given ID is no longer active, then it shall NOT be considered a protocol violation.

> This is no matter what whether the call used to be progressive or not. Otherwise the dealer needs to remember old call instances. Should that be explicitly stated or does it already follow from the text?

Hm... I don't remember exactly. WAMP expects that sequential IDs will be used by the client within the session. We definitely stated that receiving a non-sequential ID Request should be treated as a protocol violation. I suppose that's a bit different from what you are asking, but anyway decided to write. I looked through progressive chapters quickly - can't find that we explicitly stated that. So i think adding this text will be helpful for implementations.

**Comment by ecorm** [comment]
> This is no matter what whether the call used to be progressive or not. Otherwise the dealer needs to remember old call instances. Should that be explicitly stated or does it already follow from the text?

It doesn't hurt to add the rationale in the text, otherwise we'll just end up with someone asking us why we did it that way. I'll edit my proposal to include such a note. Thanks!

**Comment by ecorm** [comment]
I think we have sufficient consensus now to proceed with a PR. I'll submit one with the text I proposed above.

**Comment by ecorm** [comment]
Submitted PR #451.

**Comment by Jopie64** [comment]
> Making calls progressive can be more optimal for all sides, but most important for the dealer: as it doesn't need to instantiate a new request object in runtime, no need to generate Ids, etc. Just reuse the existing one.

I'm not against progressive call arguments as a whole, just for the non-sticky part. Your example doesn't ply for the non-sticky part right? Just for (sticky) progressive arguments.

**Comment by oberstet** [comment]
>  Otherwise the dealer needs to remember old call instances

the request ID of CALL messages of a single call with progressive call arguments MUST be the same 

this is obvious, otherwise those messages could only be logically correlated as belonging to the one call from the procedure URI being called, and thus, it would be bound to (implicit in) the transport

for a CALL with progressive call arguments, the dealer will need to keep an internal call object around until that call ends obviously

when the call ends, the dealer will still need to keep around either the internal call object, or at least the call request ID together with the flag that that call was using progressive call arguments when it still was active

this is required because of the race conditions. when the dealer receives a CALL with a request ID for a dead call, and when that dead call was using progressive arguments, it should not bail out but simply ignore

the dealer can purge this list of "old request IDs" after the maximum race time has been passed

practically: N x link latency

**Comment by Jopie64** [comment]
@oberstet 
Ah sorry, I read your comment after posting the review comments.

> when the call ends, the dealer will still need to keep around either the internal call object, or at least the call request ID together with the flag that that call was using progressive call arguments when it still was active
>
> this is required because of the race conditions. when the dealer receives a CALL with a request ID for a dead call, and when that dead call was using progressive arguments, it should not bail out but simply ignore
>
> the dealer can purge this list of "old request IDs" after the maximum race time has been passed

I'm generally against timeouts to prevent races... One has to pick a race timeout that is usually arbitrary. And once picked, everything works fine 99.9% of the time, but then a network hiccup occurs and a progressive call is received by the dealer right after the timeout and the session is closed due to wrongly flagged protocol violation.

Problems caused by data races are already hard to find because they don't occur very much. So when you try to reproduce it in the lab they tend to not occur at all. When you use a timer to 'fix' them, it's not really fixed but made even more rare and even harder to find.

Of course as everything out is a balance between cost and benefit. So what is the cost involved when calls with old (expired) ID are simply always ignored by the dealer, no matter it was a progressive call or not?

The cost is that it might not be able to flag bad behaving callers who reuse old call id's. (It's still able to find callers who use random call id's.) So what's the worst that could happen? Caller developers will find out soon enough since the dealer will not react in that case. Such a caller will probably never reach production.

And what's the cost of trying to find such clients?
The cost is introducing a race using a timeout to make it appear as little as possible and with that hard or very hard to find. Or the cost is introducing a memory leak until the caller session ends.
And even then you might not find all bad behaving callers since when such caller reuses an old progressive call ID, it's still not flagged.

My opinion is to keep it simple and just ignore the reuse of old call id's even if that means you sometimes are not able to flag bad behaving clients.

Note that when the caller didn't announce the progressive call arguments feature, the dealer can still consider usage of old call ID's a protocol violation.

**Comment by oberstet** [comment]
> One has to pick a race timeout that is usually arbitrary. And once picked, everything works fine 99.9% of the time, but then a network hiccup occurs and a progressive call is received by the dealer right after the timeout and the session is closed due to wrongly flagged protocol violation.

Yes, that's a problem as long as the network cannot guarantee a maximum latency. Which the Internet cannot obviously.

However, practically, users should run their routers with router-initiated heart-beating. This will timeout the whole connection router-side once the heartbeat timeout is hit.

In Crossbar.io, there are options that configure the underlying Autobahn transports via `autoPingInterval`, `autoPingTimeout`, `autoPingSize`, `autoPingRestartOnAnyTraffic`

If the retention time for request IDs is larger than `autoPingTimeout`, this is good.

> Of course one should always try to find a balance between cost and benefit. So what is the cost involved when calls with old (expired) ID are simply always ignored by the dealer, no matter it was a progressive call or not?

The cost is the increase in attack surface exposed, and thus additional risk (for the router).

E.g. how about: I send CALLs with made up request IDs to your router at 1k CALLs/s?

**Comment by Jopie64** [comment]
> E.g. how about: I send CALLs with made up request IDs to your router at 1k CALLs/s?

Ah... Didn't think of possible DOS attacks...

But to be fair, you still can't do that with just 'made up' request ID's. Those request IDs must have been valid and used before.
And if you use the timeout, you can still do DOS attacks when you make a progressive call and then reuse that ID. Once the timeout is almost hit just make a new progressive call with me fresh id.

Anyway, I see your point now...

**Comment by ecorm** [comment]
I think I would have preferred the simplicity of not having to temporarily remember prior progressive call request IDs, but I can live with it if it's somehow better for security. Having to periodically sweep saved request IDs is essentially time-based garbage collection.

Can the temporary storage of these request IDs be itself an attack vector? I'm thinking it's no worse than the memory required to store active progressive call requests. KSDaemon has mentioned something about limiting the number of concurrent progressive calls to curtail that possible DOS vector.

My PR is stuck in limbo until this matter is decided. Please let me know what the decision is.

> E.g. how about: I send CALLs with made up request IDs to your router at 1k CALLs/s?

Compared to legitimate calls at 1k CALL/s, this is actually less stressing to a router because there is nothing to process.

Having said that, I don't want to promote "security by lack of imagination" either.

**Comment by KSDaemon** [comment]
My position is security first. Simplicity and convenience afterward. Just because we are operating not on the level of some proprietary system that no one knows but on the level of open standards, the protocol can be used by anyone anywhere. 

At the same time, this does not mean that security must be achieved by complex measures sewn into the protocol itself. Sometimes it is much easier to solve the problem of security by additional administrative measures. And I don't see anything wrong with that. The main thing is that this is our conscious and informed choice.

**Comment by ecorm** [comment]
> At the same time, this does not mean that security must be achieved by complex measures sewn into the protocol itself. Sometimes it is much easier to solve the problem of security by additional administrative measures. And I don't see anything wrong with that. The main thing is that this is our conscious and informed choice.

So shall we make up to the router implementation to decide what to do with old request IDs in CALL messages when progressive calls are enabled? If so, this still needs to be documented in the spec.

**Comment by KSDaemon** [comment]
Let's think of the security implications of ignoring old IDs on the router side. @oberstet mentioned increasing the attack surface... 
...So the attacker may send a bunch of requests with old IDs in the same session without being disconnected with `protocol_violation` but it won't have any effect as there are no calls. in case of hijacking other sessions and thus having the ability to tamper/inject malicious data in the ongoing requests - well... If we are assuming such cases - then yes, throwing protocol_violation` makes hacker life harder.

Regarding [D]DoS attacks - well they can be possible by just issuing new calls at 1k/sec. And that is even better (from a hacker view) :)

**Comment by ecorm** [comment]
Just so that I can move on with my implementation, how about we make the strictness of CALL request ID checking "up to the implementation" (as KSDaemon suggested) and at least document there's special consideration needed for request IDs when progressive calls are supported?

**Comment by KSDaemon** [comment]
Well... Don't want to give that to the implementation side... As it may lead to a different flow experience with different router implementations. But maybe it is okay. While we are discussing and figuring that out... IDK...

**Comment by Jopie64** [comment]
Reaction to https://github.com/wamp-proto/wamp-proto/issues/451#issuecomment-1447082543

>> About that "no longer able to respond" part, does that still refer to a callee where an INVOCATION is sent to and left after that?
>> If so, I think it is wrong to use the same error code for that case. I'll explain why if confirmed that was indeed the intention.  (Cause explanation might be quite some work and if that's not the intention, that work would be for nothing)
>
> Yes, that was intention. It appears you missed/skipped over that discussion in the related issue, which started here: https://github.com/wamp-proto/wamp-proto/issues/448#issuecomment-1444239548

I didn't skip it, but at that time the sticky/non-sticky confusion was still going on... So I thought when that was resolved, this would also.

So here's why I don't think reusing `wamp.error.no_available_callee` when a callee leaves during execution of a progressive (arguments) call, is a good idea.

Currently this error code is only used when all callees replied to an `INVOCATION` with `wamp.error.unavailable`. When a callee replies with `wamp.error.unavailable`, it indicates it is busy and cannot handle the call, so it must not have had any side effects. Thats why the dealer can try the next callee.

When the callee didn't reply to the `INVOCATION` yet, but just leaves, it is unknown whether it had side-effects already. That's why the dealer will _not_ try the next callee in that case. So there could still be another untried callee left that could handle the call.

This already indicates that it is not a similar situation, and that the error should not be the same.

Furthermore, when the callee leaves during a non-progressive argument call, the dealer already responds with something. I looked in the docs. Looks like this is currently undocumented... If I remember correctly, Crossbar currently responds with `wamp.error.canceled`. Why should this be any different from progressive argument calls?

So there're actually more objections to using this error code.


1. Caller might draw wrong conclusion
   * It concludes callees are busy, hence it should retry later. While in fact there could already be another callee that's still able to handle the call.
   * It concludes call didn't have any effect. While in fact it could have had an effect already because it may have started processing.
2. When not progressive, although undocumented, another error is sent.
   * For me it's exactly the same situation. If you want to reuse an error message, why not reuse that one instead?


My suggestion would be to specify what happens when callee leaves when it's still busy handling a call. No matter whether that was a progressive or non-progressive one. I believe there's no need to document a similar situation again especially for progressive argument calls.

I hope it's clear what I mean. And maybe I overlooked something. If so, please let me know.

**Comment by KSDaemon** [comment]
@Jopie64 Good points and explanation!

**Comment by ecorm** [comment]
> Furthermore, when the callee leaves during a non-progressive argument call, the dealer already responds with something. I looked in the docs. Looks like this is currently undocumented... If I remember correctly, Crossbar currently responds with wamp.error.canceled. Why should this be any different from progressive argument calls?

If you're referring to the *Progressive Call Cancellation* subsection, it only specifies the behavior for the **Caller** leaving, not the **Callee**. It doesn't mention anything about error URIs.

If I missed it, please point me to where it specifies behavior for the **Callee** leaving in the spec.

> This already indicates that it is not a similar situation, and that the error should not be the same.

Okay, you've convinced me, but I don't agree about reusing `wamp.error.cancelled` for the very same reason: the client can't distinguish between the Callee bailing out or an CANCEL message being processed.

So, to me, the question remains what should be the error URI for a Callee leaving after an INVOCATION it sent (without the Callee responding with a RESULT or ERROR)? How about `wamp.error.callee_left`? Short, simple, self-explanatory.

**Comment by ecorm** [comment]
> That's why the dealer will not try the next callee in that case. So there could still be another untried callee left that could handle the call.

Where in the spec is this behavior mandated? If it's not, then it should be, one way or another.

If the wish the spec had permalinks at the subsection level. It would help in settling these sorts of debates.

**Comment by ecorm** [comment]
> How about wamp.error.callee_left? Short, simple, self-explanatory.

Or perhaps `wamp.error.callee_left_before_responding`? More verbose, but clearer.

**Comment by KSDaemon** [comment]
> How about `wamp.error.callee_left`? Short, simple, self-explanatory.
Or perhaps `wamp.error.callee_left_without_responding`? More verbose, but clearer.

I like `wamp.error.callee_left` or `wamp.error.callee_gone`. No need in `_without_responding`. Because if callee responded and then leaved — calller won't event know that :)

**Comment by Jopie64** [comment]
If `wamp.error.canceled` is not good enough then I also vote for `wamp.error.callee_left`.

And indeed without the `without_responding` part because the callee could already have responded with some progressive results.

Thing what scares me though is that the Crossbar router is already out there responding with `wamp.error.canceled` in this case (I think). So this would be a breaking change for Crossbar...
I guess that's what you get when not specifying behavior...

@oberstet what do you think? Got this situation before? How to handle that? Configuration setting to use old behavior? Make a note in the release notes of breaking changes? Something else?

**Comment by Jopie64** [comment]
@ecorm 
> If you're referring to the Progressive Call Cancellation subsection, it only specifies the behavior for the Caller leaving, not the Callee. It doesn't mention anything about error URIs.

I was referring to general RPC behaviour. And indeed as far as I can see it doesn't say anything about the callee leaving during a call (invocation). But that's also what we're discussing now 😊

**Comment by oberstet** [comment]
@Jopie64 I regret having merged a feature into the spec without having an own implementation;) I will not change the error URIs in crossbar without an implementation of progressive call arguments

**Comment by ecorm** [comment]
> Thing what scares me though is that the Crossbar router is already out there responding with wamp.error.canceled in this case (I think). So this would be a breaking change for Crossbar...

If not using `wamp.error.canceled` for this use case would lead to problems due to breaking existing behavior, then I'm not opposed to adopting it.

> I guess that's what you get when not specifying behavior...

I'll be honest and say that I don't like these situations where the spec is forced to conform to "what Crossbar does" without any deliberation from the WAMP community. I also understand that having a working model helps to advance the elaboration of the spec.

> I regret having merged a feature into the spec without having an own implementation;) I will not change the error URIs in crossbar without an implementation of progressive call arguments

But the "callee leaving before responding" use case existed way before the progressive call <whatever> feature! Race conditions were never really analyzed, as far as I can tell.

**Comment by Jopie64** [comment]
@oberstet 

It's not necessarily about progressive calls. It's about the missing specification what happens when a callee leaves when it received an invocation but didn't respond with a final YIELD or ERROR yet. In that case Crossbar dealer respons to the caller, I think with wamp.error.cancelled. But as said this is not documented.

We do want to add this behaviour to the specs but with a different error URI. See discussion above.

**Comment by ecorm** [comment]
@oberstet I'll repeat Jopie64's diagram here to refresh your memory:

```mermaid
sequenceDiagram
    participant Caller
    participant Dealer
    participant Callee

    Caller->>Dealer: CALL (progress=false)
    Dealer->>Callee: INVOCATION
    Callee->>Dealer: GOODBYE
    Dealer->>Caller: CALL ERROR (cancelled)
```

From the Caller's perspective, it cannot distinguish between the above scenario and a Caller-initiated CANCEL upon receiving a `wamp.error.cancelled` error.

**Comment by ecorm** [comment]
> Me: From the Caller's perspective, it cannot distinguish between the above scenario and a Caller-initiated CANCEL upon receiving a wamp.error.cancelled error.

Thinking about this some more... does this really matter? Whether it's from responding to an INTERRUPT message, or from the Callee leaving, the desired effect is the same from the Caller's point of view: the pending call was terminated.

If the Caller receives a `wamp.error.cancelled` out of the blue without having issued a CANCELLED, then it can deduce that it was due to the Callee leaving before responding.

I'm now okay with adopting `wamp.error.cancelled` for the "callee leaving before responding" use case. This has the advantage of not breaking Crossbar's existing behavior, and not having to introduce a new error URI.

**Comment by ecorm** [comment]
> Me: I'll be honest and say that I don't like these situations where the spec is forced to conform to "what Crossbar does" without any deliberation from the WAMP community. I also understand that having a working model helps to advance the elaboration of the spec.

I should add that I'm not blaming anyone. It is what it is. I understand this is a volunteer effort for some, and that people need sleep and have healthy work/life balance. Keeping this massive standard in sync with implementations is a lot of work!

**Comment by KSDaemon** [comment]
> I'm now okay with adopting `wamp.error.cancelled` for the "callee leaving before responding" use case. This has the advantage of not breaking Crossbar's existing behavior, and not having to introduce a new error URI.

Agree! 💯 

> I'll be honest and say that I don't like these situations where the spec is forced to conform to "what Crossbar does" without any deliberation from the WAMP community. I also understand that having a working model helps to advance the elaboration of the spec.

And here I agree too. Of course, we do not live in an ideal world. So certainly we can be in situations when implementations are out of sync with the spec. But we improve and evolve the spec, update the implementations, and vice versa. I don't see any problems here.

For example, we removed `sticky` option. Nexus already implemented progressive call invocations and described this option. So what? We shall keep it forever? No. My response — is just to follow semver and update the major version in this case and explicitly state in the documentation about breaking change and how to upgrade (if it is possible). I do like this all the time in wampy.js for example. And that's it.

**Comment by Jopie64** [comment]
@ecorm
> If the Caller receives a wamp.error.cancelled out of the blue without having issued a CANCELLED, then it can deduce that it was due to the Callee leaving before responding.

That's what I thought at first. But when I was about to reply something similar to your first complaint about reusing `wamp.error.canceled`, I also thought, are there situations where the caller could be wrong? And there we get those dreaded races again...

```mermaid
sequenceDiagram
    participant Caller
    participant Network
    participant Dealer
    participant Callee

    Caller->>+Network: CALL
    Network->>-Dealer: CALL
    Dealer->>Callee: INVOCATION
    Caller->>+Network: CANCEL
    Callee->>Dealer: GOODBYE
    Dealer->>+Network: wamp.error.canceled
    Network->>-Caller: wamp.error.canceled
    note right of Caller: Concludes canceled<br>received due to<br>CANCEL
    Network->>-Dealer: CANCEL
    note right of Dealer: Race detected<br>Call already ended<br>Ignore

```

`wamp.error.canceled` received due to leave. But caller concludes it was because of `CANCEL`.

Still the question: Does this matter? The effect is the same: call is canceled. Only the cause is different...

> Thinking about this some more... does this really matter? Whether it's from responding to an INTERRUPT message, or from the Callee leaving, the desired effect is the same from the Caller's point of view: the pending call was terminated.

So this still stands.

**Comment by ecorm** [comment]
Good analysis Jopie64, thanks!

Analyzing all the possible race conditions and "abnormal" situations in routed RPCs would make a good computer science student project. Could we hire ChatGPT to do it? :grin:

**Comment by KSDaemon** [comment]
Ough yeah....

So back again to question: what is bad if caller receives CANCEL for the non-existent call and just ignores it?

**Comment by Jopie64** [comment]
@KSDaemon 
> So back again to question: what is bad if caller receives CANCEL for the non-existent call and just ignores it?

Caller never receives `CANCEL`, it only sends it. It may receive `wamp.error.canceled` error when it sent `CANCEL` with `killnowait: true` due to above race. But just read the docs, even that is expected: Caller always receives `ERROR` from dealer when sending `CANCEL`.

**Comment by KSDaemon** [comment]
@Jopie64 You're right. CANCEL does not expect feedback.

**Comment by KSDaemon** [comment]
Need to think about it more, but in the first place I don't see any bad behaviors omitting the second ERROR/CANCEL/etc messages. I definitely against making these cases `protocol_violations`

**Comment by ecorm** [comment]
> I definitely against making these cases protocol_violations

Agreed, unless we can formally prove that they cannot happen by "well-behaving" peers, with that "good behavior" being explicit in the spec.

**Comment by KSDaemon** [comment]
Are all the points addressed here resolved? So we can close it?

**Comment by ecorm** [comment]
Spawned sub-issues | Resolved by
------------------ | -----------
#450 | #460
#476 | #478
#482 | #491
#494 | #495
